Tasks/NodeTool/nodetool.ts

```diff
@@ -27,149 +32,136 @@ async function run() {
-
-        let downloadPath: string = await toolLib.downloadTool(downloadUrl);
-
-        //
-        // Extract the tar and install it into the local tool cache
-        //
-        let extPath: string;
-        if (osPlat == 'win32') {
-            taskLib.assertAgent('2.115.0');
-            extPath = taskLib.getVariable('Agent.TempDirectory');
-            if (!extPath) {
-                throw new Error('Expected Agent.TempDirectory to be set');
-            }
-
-            extPath = path.join(extPath, 'n'); // use as short a path as possible due to nested node_modules folders
-            extPath = await toolLib.extract7z(downloadPath, extPath);
-        }
-        else {
-            extPath = await toolLib.extractTar(downloadPath);
+        if (!toolPath) {
+            // download, extract, cache
+            toolPath = await download(version);
         }
-
-        // node extracts with a root folder that matches the fileName downloaded
-        let toolRoot = path.join(extPath, fileName);
-        
-        toolLib.cacheDir(toolRoot, 'node', version);
     }
 
-    console.log('using version: ' + version);
-
     //
     // a tool installer initimately knows details about the layout of that tool
     // for example, node binary is in the bin folder after the extract on Mac/Linux.
     // layouts could change by version, by platform etc... but that's the tool installers job
     //
-    let toolPath: string = toolLib.findLocalTool('node', version);
     if (osPlat != 'win32') {
         toolPath = path.join(toolPath, 'bin');
     }
 
-    console.log('using tool path: ' + toolPath);
-
     //
     // prepend the tools path. instructs the agent to prepend for future tasks
     //
     toolLib.prependPath(toolPath);
-    console.log();
+}
+
+async function queryLatestMatch(versionSpec: string): Promise<string> {
+    // node offers a json list of versions
+    let dataFileName: string;
+    switch (osPlat) {
+        case "linux": dataFileName = "linux-" + osArch; break;
+        case "darwin": dataFileName = "osx-" + osArch + '-tar'; break;
+        case "win32": dataFileName = "win-" + osArch + '-exe'; break;
+        default: throw new Error(`Unexpected OS '${osPlat}'`);
+    }
+
+    let versions: string[] = [];
+    let dataUrl = "https://nodejs.org/dist/index.json";
+    let rest: restm.RestClient = new restm.RestClient('tool-sample');
+    let nodeVersions: INodeVersion[] = (await rest.get<INodeVersion[]>(dataUrl)).result;
+    nodeVersions.forEach((nodeVersion:INodeVersion) => {
+        // ensure this version supports your os and platform
+        if (nodeVersion.files.indexOf(dataFileName) >= 0) {
+            versions.push(nodeVersion.version);
+        }
+    });
+
+    // get the latest version that matches the version spec
+    let version: string = toolLib.evaluateVersions(versions, versionSpec);
+    return version;
+}
+
+async function download(version: string): Promise<string> {
```

Comment:
Also, adding "v" prefix into the URL format strings was the other significant change here. And added `version = toolLib.cleanVersion(version);` so the cleaned URL can be appended. Again, this is required to fix a bug today where the user enters the version without the v prefix.